DSBSparse: Addressing some performance issues

vetschn opened the pull request:
Several things are still done naively in the DSBSparse matrices.

- [x] Caching mechanism for the DSBCOO block-wise access. Since we already have the rowptr map in the DSBCSR, there are no changes needed there. Also Mentioned in #4.
- [x] Return blocks in some sort of sparse format. Closes #4.
- [x] #43 
- [ ] ~~Better tests for arithmetic~~

vetschn committed:
Caching of block-slices.

I added a caching mechanism for the DSBCOO block-wise access. Since we already have the rowptr map in the DSBCSR, there are no changes needed there.

vetschn committed:
Ability to return sparse blocks in DSBSparse matrices.

I also added some tests that should make clear how to use the returned arrays.

vetschn committed:
First steps towards generalized, vectorized item-assignment in DSBSparse.

Similar to what i did for the _set_block/_get_block methods, the __setitem__/__getitem__ methods now wrap _set_items/_get_items. These functions can set a list of elements in a vectorized way. This also works on a specific part of the stack and both when distributed through stack and nnz.

vetschn commented:
So i added these vectorized getters and setters now in the `DSBCOO` class (`DSBCSR` is next). The behavior is like this:

- If we are in the **`"stack"`** distribution state, you'll get an array of the expected shape (padded with zeros where requested elements are not in the matrix). So 
  ```python
  inds = np.arange(10)
  values = dsbcoo[inds, inds]
  ```
  will get `values.shape[-1] == 10` with maybe some zeros in there, where `(ind[i], ind[i])` is not in the matrix. You can also set items with the same logic:
  ```python
  inds = np.arange(10)
  values = np.ones(10)
  dsbcoo[inds, inds] = values
  ```
  This will only change those elements that are actually in the matrix, the sparsity pattern is immutable as always.

- If we are in the **`"nnz"`** distribution, the same logic should apply as before this change. If you request an element that is not in the matrix, you'll get an `IndexError` (because we cannot know which rank is supposed to hold an element that is not in the matrix). 

  If you request elements that are all in the matrix but distributed accross ranks, each rank now only returns the requested elements that it holds. On the ranks where there are no elements on the requested inds, you'll get an empty array (of the correct stack shape).
  ```python
  inds = np.arange(10)
  values = dsbcoo[inds, inds]
  ```
  could be `values.shape[-1] == 7 if comm.rank == 0`, `values.shape[-1] == 3 if comm.rank == 1`, and `values.shape[-1] == 0 if comm.rank == 2`.

  Setting elements in this distribution state works analogously:
    ```python
  inds = np.arange(10)
  values = np.ones(10)
  dsbcoo[inds, inds] = values
  ```
  Each rank only sets those elements of the input, which it owns. If we are accessing elements that are not in the matrix nothing happens and the sparsity remains unchanged.

Note that everything here also works when accessing only a part of the stack, so random things like
```python
inds = np.arange(10)
values = dsbcoo.stack[:3][inds, inds]
dsbcoo.stack[[2, 4, 6]][inds, inds] = 1.5 * values
```
should work as well in both distribution states.

vetschn committed:
Abstract away the stack index ellipsis replacement.

vetschn committed:
First attempt towards working CSR items setters/getters (everything is broken)

vetschn committed:
Get rid of commented out code

vetschn committed:
Implemented vectorized item getter/setter in DSBCSR

vetschn committed:
Small bugfix

vetschn committed:
Refactor the resolution of xp/sparse backends.

Determining xp and sparse backends now happens at the top-level init. I removed unnecessary transfers, everything stays on GPU except when it computes eigenvalues.

I also adjusted tests to make sense with these changes. All tests passing on the GPU & CPU.

vetschn committed:
More GPU/CPU agnostic stuff

vetschn committed:
Add tests for distributed_load function with NPY and NPZ formats

vetschn committed:
Use correct tmp_path fixture

vincent-maillou commented on src/qttools/__init__.py:
I like that, didn't think about it but find it very clean.

vincent-maillou commented on src/qttools/utils/mpi_utils.py:
```diff
@@ -2,10 +2,12 @@
 from pathlib import Path
 
-import numpy as np
+import scipy.sparse as sps
```
Why do you still need scipy.sparse explicitly here? And don't use the sparse from your __init__?

vincent-maillou replied:
EDIT: I guess it's to use the load_npz from host side, why not using: https://docs.cupy.dev/en/latest/reference/generated/cupy.load.html in the case of xp = cp?

vincent-maillou commented on tests/obc/test_spectral.py:
```diff
@@ -101,12 +100,12 @@ def test_memoizer(
     a_ji, a_ii, a_ij = _make_periodic(a_xx, block_sections)
     x_ii = spectral(a_ii=a_ii, a_ij=a_ij, a_ji=a_ji, contact=contact)
-    assert np.allclose(x_ii, xp.linalg.inv(a_ii - a_ji @ x_ii @ a_ij), atol=1e-5)
+    assert xp.allclose(x_ii, xp.linalg.inv(a_ii - a_ji @ x_ii @ a_ij), atol=1e-5)
```
I'm not a huge fan of having the tolerances hard-coded here, I guess we  should have all solvers convergence tol somewhere? Anyway not a big deal

vetschn replied:
Definitely not a fan either. I think we talked about this already at some point but to test the Spectral OBC solver and the NEVP we should somehow construct matrices with hand-picked eigenvalues. That could maybe also solve this?

Alternatively, cranking up the refinement iterations always helps :sweat_smile:

vincent-maillou replied:
Yes, maybe open an issue for the eigenvalues solvers tests, I foresee @almaeder being very suited for this issue :smile:

vetschn replied:
See #51

vincent-maillou reviewed (approved):
I'm not sure of the quality of my review as my brain stopped working a while ago today. But anyway: LGTM :1st_place_medal:

vetschn commented on src/qttools/datastructures/dsbcoo.py:
```diff
@@ -147,29 +248,32 @@ def _get_block(self, stack_index: tuple, row: int, col: int) -> ArrayLike:
+        if not self.return_dense:
+            if block_slice == slice(None):
+                # No data in this block, return an empty block.
+                return xp.empty(0), xp.empty(0), xp.empty(data_stack.shape[:-1] + (0,))
```
This is the code that handles returning an empty "sparse" matrix, so empty `rows`, `cols` and `data`.

vincent-maillou replied:
make sense, thanks

vetschn commented on src/qttools/datastructures/dsbcsr.py:
```diff
@@ -55,98 +54,175 @@ def __init__(
+            rr = rows[mask] - self.block_offsets[brow]
+            cc = cols[mask]
+
+            # TODO: This could perhaps be done in an efficient way.
```
See #50

vincent-maillou replied:
amazing! thks

vetschn committed:
Fix for cupyx.scipy.sparse does not implement LIL matrices.

Had to rewrite the logic to add a sparse matrix to the DSBSparse matrices. Absolute pain.

vetschn committed:
Abstract away the in-place addition/subtraction/multiplication

vetschn committed:
Realized my life is for once easier than i thought.
